### ISSUE TITLE
Prior art: API design

### ISSUE THREAD

**Issue by curiousleo** [lehins/random#13]
# Prior art in "standard" libraries

## Go

Golang has two APIs, one for non-crypto PRNGs and one for cryptographically random bytes:

- https://golang.org/pkg/math/rand/
- https://golang.org/pkg/crypto/rand/

## Rust

Rust has a family of packages for PRNGs. The basic interfaces are defined in `rand_core`.

- https://rust-random.github.io/rand/rand_core/index.html
- https://rust-random.github.io/book/crates.html

## C++

The C++ standard defines a set of highly templated classes which together form the PRNG API.

- https://en.cppreference.com/w/cpp/numeric/random

O'Neill makes the point that this API is flexible, but hard to use correctly, and proposes an alternative interface. IMO it's a nice API that provides safe defaults but is still very flexible for those who want flexibility.

- https://www.pcg-random.org/posts/ease-of-use-without-loss-of-power.html
- https://gist.github.com/imneme/540829265469e673d045

## numpy.random

[Inspired by O'Neill](https://github.com/numpy/numpy/blob/acba24414487e0e171bedbc8d1ae3cfc63f4b2c7/numpy/random/_bit_generator.pyx#L4-L5).

- https://numpy.org/devdocs/reference/random/parallel.html

# Prior art in Haskell libraries

## `splitmix`

Pure, splittable SplitMix implementation.

- https://hackage.haskell.org/package/splitmix

## `mwc-random`

Provides a monadic generator interface for Marsaglia's MWC256 (a.k.a. MWC8222) multiply-with-carry generator, a way to access system randomness, and a way to draw from a number of random distributions. Does not provide a way to split/spawn generators or seeds.

- http://hackage.haskell.org/package/mwc-random
- http://hackage.haskell.org/package/mwc-random-0.14.0.0/docs/System-Random-MWC-SeedSource.html
- http://hackage.haskell.org/package/mwc-random-0.14.0.0/docs/System-Random-MWC-Distributions.html

## `random-source`

Defines the `MonadRandom` typeclass, an interface for random number generation. One aim of our effort here should be to make `random-source` unnecessary (see https://github.com/idontgetoutmuch/random/issues/4#issuecomment-592950462).

- https://hackage.haskell.org/package/random-source

## `random-fu`

Provides a way to draw from a number of random distributions. The randomness source is pluggable AFAICT.

- https://hackage.haskell.org/package/random-fu

## `tf-random`

Implements a splittable PRNG implementing the old `random`'s `RandomGen`, provides alternative interfaces, and a way to access system randomness. Based on ThreeFish.

- https://hackage.haskell.org/package/tf-random

# More prior art

- JAX: [API](https://jax.readthedocs.io/en/latest/jax.random.html), [design doc](https://github.com/google/jax/blob/master/design_notes/prng.md), [gotchas](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#JAX-PRNG)

# Common concepts

The API split between randomness sources and random distributions is apparent in every API I looked at:

- Randomness source: [`Source`](https://golang.org/pkg/math/rand/#Source) in Go, [`RngCore`](https://rust-random.github.io/rand/rand_core/trait.RngCore.html) and [`Rng`](https://rust-random.github.io/rand/rand/trait.Rng.html) (automatically derived from `RngCore`) in Rust, [`UniformRandomBitGenerator`](https://en.cppreference.com/w/cpp/named_req/UniformRandomBitGenerator) in C++, [`BitGenerator`](https://numpy.org/devdocs/reference/random/bit_generators/generated/numpy.random.BitGenerator.html#numpy.random.BitGenerator) in numpy. These can be given a seed, and they generate random bits/bytes/words.
- Drawing from a distribution: [`Rand`](https://golang.org/pkg/math/rand/#Rand) in Go, implementors of the [`Distribution`](https://rust-random.github.io/rand/rand_distr/trait.Distribution.html) trait in Rust, components implementing [`RandomNumberDistribution`](https://en.cppreference.com/w/cpp/named_req/RandomNumberDistribution) in C++, `Generator` in numpy. These use a randomness source to draw from a distribution.

Note that drawing from a distribution is what [`random-fu`](https://hackage.haskell.org/package/random-fu) exists for in the Haskell world. I believe it is worth understanding if and what changes would be required to make `random-fu` work with the new `random`. Apart from that, it seems to cover the issue of drawing from a distribution well, so we can concentrate on building a solid API for randomness sources and seeding here.

### Seeding

Seeding is handled a bit differently from API to API.

- Rust's `rand` has the [`SeedableRng`](https://rust-random.github.io/rand/rand_core/trait.SeedableRng.html) trait, whose associated `Seed` type must be convertible to byte slices.
  RNGs can be seeded directly, from an integer value (which is mixed to increase entropy), or from another RNG (which appropriate warnings in the docs).
- C++ has [`random_device`](https://en.cppreference.com/w/cpp/numeric/random/random_device) to get randomness from the system and [`seed_seq`](https://en.cppreference.com/w/cpp/numeric/random/seed_seq) to generate good seeds from low-entropy inputs, which can be used to instantiate randum number "engines" like the [Mersenne Twister](https://en.cppreference.com/w/cpp/numeric/random/mersenne_twister_engine/mersenne_twister_engine).
- numpy works similarly to C++, where the relevant class is called [`SeedSequence`](https://numpy.org/devdocs/reference/random/bit_generators/generated/numpy.random.SeedSequence.html#numpy.random.SeedSequence), but with a nicer API and fewer footguns.
- Go has the [`Rand.Seed`](https://golang.org/pkg/math/rand/#Rand.Seed) function.

**Comment by idontgetoutmuch** [comment]
An interesting discussion https://github.com/JuliaLang/julia/issues/27614 not too dissimilar to the one we've been having.
